### PR TITLE
chore: refactor for typing, proper version checks, and more

### PR DESCRIPTION
**Pull request by omercnet**
## Related Issues

Fixes https://github.com/descope/etc/issues/4039

## Description
Expanding support for python 3.11
Adding typing annotations everywhere, fix bugs that it uncovered
Use tox for automated testing
Add tests for all platforms
limiting support for 3.7 to test only

## Must

- [x] Tests

### CODE REVIEW

File: descope/auth.py [hunk 1]
```diff
@@ -614,21 +617,27 @@ def refresh_session(
...
 
         try:
             return self.validate_session(session_token, audience)
         except Exception:
             # Session is invalid - try to refresh it
+            if not refresh_token:
+                raise AuthException(
+                    400,
+                    ERROR_TYPE_INVALID_TOKEN,
+                    "Session token is missing",
```
[the remark this answers is not in the document]
Comment (omercnet) [reply descope/auth.py:639]:
```suggestion
                    "Refresh token is missing",
```

File: descope/auth.py [hunk 2]
```diff
@@ -214,7 +225,7 @@ def compose_url(base: str, method: DeliveryMethod) -> str:
         return f"{base}/{suffix}"
 
     @staticmethod
-    def get_login_id_by_method(method: DeliveryMethod, user: dict) -> Tuple[str, str]:
+    def get_login_id_by_method(method: DeliveryMethod, user: dict) -> tuple[str, str]:
```
[the remark this answers is not in the document]
Comment (omercnet) [reply descope/auth.py:228]: annotations replace the typing library 
see the top line `from __future__ import annotations` that provides backwards support

[the remark this answers is not in the document]
Comment (omercnet) [reply descope/auth.py:228]: but this reminded me to add py.typed so consuming packages know to utilize the annotations

File: descope/auth.py [hunk 3]
```diff
@@ -380,25 +391,25 @@ def _fetch_public_keys(self) -> None:
     def adjust_properties(self, jwt_response: dict, user_jwt: bool):
         # Save permissions, roles and tenants info from Session token or from refresh token on the json top level
         if SESSION_TOKEN_NAME in jwt_response:
-            jwt_response["permissions"] = jwt_response.get(SESSION_TOKEN_NAME).get(
+            jwt_response["permissions"] = jwt_response.get(SESSION_TOKEN_NAME, {}).get(
```
[the remark this answers is not in the document]
Comment (omercnet) [reply descope/auth.py:394]: a type error without context will be even harder to debug, let's either validate the contents before, or catch that type error and raise a meaningful exception on what went wrong

**Review by guyp-descope — APPROVED** [review]

**Pull request merged**